compiler/modules/front.syntax/tests/posTree.bal

```diff
@@ -0,0 +1,108 @@
+type TerminalSyntaxNode TerminalSyntaxAstNode|FixedSyntaxNode|IdentifierSyntaxNode;
+type AstSyntaxNode TerminalSyntaxAstNode|NonTerminalSyntaxNode;
+type SyntaxNode NonTerminalSyntaxNode|TerminalSyntaxNode;
+
+type TerminalSyntaxAstNode record {|
+    AstNode astNode;
+    string? token;  // if non-nil expect the token to be this
+|};
+
+type NonTerminalSyntaxNode record {|
+    AstNode astNode;
+    SyntaxNode[] childNodes;
+|};
+
+type RootSyntaxNode record {|
+    ModulePart part;
+    SyntaxNode[] childNodes;
+|};
+
+// represents keywords and fixed tokens
+type FixedSyntaxNode record {|
+    FixedToken token;
+    Position? pos = ();
+|};
+
+// represents identifiers
+type IdentifierSyntaxNode record {|
+    string name;
+    Position? pos;
+|};
+
+type AstNode record {|
+   *PositionFields;
+   any...;
+|};
+
+
+function validateModulePart(ModulePart part) {
+    RootSyntaxNode root = buildTree(part);
+    validateSyntaxNode(root, part.file.filename());
+}
+
+function buildTree(ModulePart part) returns RootSyntaxNode {
+    SyntaxNode[] importDecls = from ImportDecl decl in part.importDecls select syntaxNodeFromImportDecl(decl);
+    SyntaxNode[] moduleLevelDefns = from ModuleLevelDefn defn in part.defns select syntaxNodeFromModuleLevelDefn(defn);
+    SyntaxNode[] childNodes = [];
+    childNodes.push(...importDecls);
+    childNodes.push(...moduleLevelDefns);
+    return { childNodes, part };
+}
+
+function validateSyntaxNode(SyntaxNode|RootSyntaxNode node, string filename) {
```

Comment:
Use the normal error and location handling machinery. Don't pass a filename. Make validate return a distinct error type.

Everything should take a Tokenizer as the first parameter (from which you can get a location)

Reply:
Fixed. I added the code related to errors in to `posTree.bal` instead to `err` module since these errors won't be triggered by the compiler normally (unlike other errors there). Is this the right place to put them